### PR TITLE
feat: add options to expose ports

### LINKED ISSUE #282: Expose ports without rebuilding utils image
**Is your feature request related to a problem? Please describe.**
Currently we need to build a new utils image to expose some other ports. It's not convenient for developers.

**Describe the solution you'd like**
I sugguest adding option `--expose-ports` in xud.sh. The syntax could be

```
--expose-ports geth/8545
--expose-ports geth/8545,bitcoind/8832
--expose-ports geth/192.168.1.100:8545:8545
--expose-ports geth/8545/tcp
```

The value of `--expose-ports` is comma separated and every part follows `<node>/<published-port>` which `published-port` follows the same pattern as [docker `-p` option](https://docs.docker.com/config/containers/container-networking/#published-ports)

**Describe alternatives you've considered**
I think add options in &lt;network&gt;.conf is not good. I sugguest all options for developers should be limited in xud.sh command line options.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by reliveyy**
This commit adds "--expose-ports" command-line option and several config file fields to enable ports exposure.

Closes #282

### CODE REVIEW

**Comment by erkarl** [comment]
```
➜  ~ bash xud.sh -b feat/expose-ports
1) Simnet
2) Testnet
3) Mainnet
Please choose the network: 2
Pulling image exchangeunion/utils:latest__feat-expose-ports
🚀 Launching testnet environment
❌ Failed to launch testnet environment. For more details, see /home/ar/.xud-docker/testnet/logs/testnet-2020-03-17-14-46-07.log
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/urllib/request.py", line 1319, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/local/lib/python3.8/http/client.py", line 1230, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.8/http/client.py", line 1276, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.8/http/client.py", line 1225, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.8/http/client.py", line 1004, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.8/http/client.py", line 944, in send
    self.connect()
  File "/usr/local/lib/python3.8/http/client.py", line 1392, in connect
    super().connect()
  File "/usr/local/lib/python3.8/http/client.py", line 915, in connect
    self.sock = self._create_connection(
  File "/usr/local/lib/python3.8/socket.py", line 787, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/usr/local/lib/python3.8/socket.py", line 918, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -3] Try again

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/launcher-1.0.0-py3.8.egg/launcher/config/config.py", line 731, in get_nodes_json
    r = urlopen(f"https://raw.githubusercontent.com/ExchangeUnion/xud-docker/{self.branch}/nodes.json")
  File "/usr/local/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/lib/python3.8/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/usr/local/lib/python3.8/urllib/request.py", line 542, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/usr/local/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/local/lib/python3.8/urllib/request.py", line 1362, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/usr/local/lib/python3.8/urllib/request.py", line 1322, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno -3] Try again>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/launcher-1.0.0-py3.8.egg/launcher/__init__.py", line 130, in launch
    cfg.parse()
  File "/usr/local/lib/python3.8/site-packages/launcher-1.0.0-py3.8.egg/launcher/config/config.py", line 379, in parse
    node_json = self.get_nodes_json()
  File "/usr/local/lib/python3.8/site-packages/launcher-1.0.0-py3.8.egg/launcher/config/config.py", line 734, in get_nodes_json
    raise NodesJsonMissing()
launcher.config.config.NodesJsonMissing
```

**Comment by reliveyy** [comment]
> File "/usr/local/lib/python3.8/socket.py", line 918, in getaddrinfo
> for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
> socket.gaierror: [Errno -3] Try again

This `gaierror` indicates a temporary network issue when downloading nodes.json from GitHub. I think it is unrelated to the PR. @erkarl

**Comment by erkarl** [comment]
After editing the config file to:
```
[xud]
# 18885 - P2P port
# 18886 - gRPC port
# 18887 - raiden resolver port
# 18080 - webproxy port
expose-ports = [18885, 18886, 18887, 18080]
```

It looks like the gRPC port is not exposed:
```
➜  xud git:(master) ./bin/xucli --tlscertpath=/home/ar/.xud-docker/testnet/data/xud/tls.cert --rpcport 18886 getinfo
could not connect to xud at localhost:18886, is xud running?
```

**Comment by reliveyy** [comment]
@erkarl Did you restart the xud.sh? And what is the result of `docker inspect testnet_xud_1`?

File: images/utils/launcher/config/testnet.conf [hunk 1]
```diff
@@ -41,10 +54,37 @@
...
 #rpc-host = "127.0.0.1"
 #rpc-port = 8545
 #infura-project-id = ""
 #infura-project-secret = ""
+
+[lndbtc]
+# 19735 - P2P port
+# 20009 - gRPC port
+# 20010 - REST port
+#expose-ports = [19735, 20009, 20010]
```
Comment (sangaman) [inline images/utils/launcher/config/testnet.conf]: Should we change the rpc port to 10009 until we fix the mappings? It's the only way I've had success with this as discussed.

File: images/utils/launcher/config/testnet.conf [hunk 2]
```diff
@@ -41,10 +54,37 @@
...
+[raiden]
+# 15001 - raiden API port
+#expose-ports = [5001]
+
+[xud]
+# 18885 - P2P port
+# 18886 - gRPC port
+# 18887 - raiden resolver port
+# 18080 - webproxy port
+#expose-ports = [18885, 18886, 18887, 18080]
```
Comment (sangaman) [inline images/utils/launcher/config/testnet.conf]: nit: new line EOF?

File: images/utils/launcher/config/testnet.conf (hunk 1 above)
Comment (reliveyy) [reply images/utils/launcher/config/testnet.conf]: Fixed.

File: images/utils/launcher/config/testnet.conf (hunk 2 above)
Comment (reliveyy) [reply images/utils/launcher/config/testnet.conf]: Fixed.

File: images/utils/launcher/check_wallets.py [hunk 3]
```diff
@@ -100,19 +98,29 @@ def check_restore_dir_files(self, restore_dir):
...
         exit_code = os.system(f"grep -q backup-dir {config_file} >/dev/null 2>&1")
 
         if exit_code == 0:
-            if not self.config.backup_dir:
-                raise NetworkConfigFileSyntaxError(f"The \"backup-dir\" option is in a wrong section of {network}.conf. Please check your configuration file.")
-            os.system(f"sed -Ei 's/^.*backup-dir = .*$/backup-dir = {backup_dir}' {config_file}")
+            os.system(f"sed -Ei 's|backup-dir = .*$|backup-dir = \"{backup_dir}\"|' {config_file}")
+            line = check_output(f"grep backup-dir {config_file}", shell=True).decode().splitlines()[0].strip()
+            if line.startswith("#"):
+                # uncomment backup-dir line
```
[the remark this answers is not in the document]
Comment (reliveyy) [reply images/utils/launcher/check_wallets.py]: It's just an explanation for the purpose of the code block.

File: images/utils/launcher/config/mainnet.conf [hunk 4]
```diff
@@ -41,10 +54,36 @@
...
+# 9735 - P2P port
+# 10009 - gRPC port
+# 10010 - REST port
+#expose-ports = [9735, 10009, 10010]
+
+[lndltc]
+# 10735 - P2P port
+# 10009 - gRPC port
+# 10010 - REST port
+#expose-ports = [10735, "20009:10009", "20010:10010"]
```
Comment (sangaman) [inline images/utils/launcher/config/mainnet.conf]: I believe this doesn't work due to an `Not a homogeneous array (line 54 column 1 char 1687)` error, I needed to make all values in the array strings for it to work. I haven't tried yet with an array of all numbers but I suspect we should make them strings as well, and maybe follow the `exposedport:containerport` format for all `export-ports` values in all sample config files.

Comment (erkarl) [reply images/utils/launcher/config/mainnet.conf]: Nice catch, sorry I missed this. I tested it with old config - all strings worked.

Comment (reliveyy) [reply images/utils/launcher/config/mainnet.conf]: Thanks for point this out. It is a trap that users could easily fall into so I quotes all values in expose-ports option for safe.

Comment (reliveyy) [reply images/utils/launcher/config/mainnet.conf]: @kilrau Should we add some instructions about  the advanced expose-ports syntax in the sample config file?

Comment (kilrau) [reply images/utils/launcher/config/mainnet.conf]: We cant have instructions, nor ""s and mallings in the cobfig file. Converting to string and mapping has to be handled by xud-docker code. The config file should look exactly like in https://github.com/ExchangeUnion/xud-docker/issues/282#issuecomment-598378806

Comment (erkarl) [reply images/utils/launcher/config/mainnet.conf]: I'm confused. Why is `expose-ports = ["10735", "20009:10009", "20010:10010"]` not a valid option?

Comment (reliveyy) [reply images/utils/launcher/config/mainnet.conf]: @kilrau So do you suggest to change
```
#expose-ports = [10735, "20009:10009", "20010:10010"]
```
to
```
#expose-ports = [10735, 10009, 10010]
```
like it described in the https://github.com/ExchangeUnion/xud-docker/issues/282#issuecomment-598378806.

But you could notice there are still some changes I've made. In the #282 comment for lndbtc it should be:
```
[lndltc]
# 10735 - P2P port
# 11009 - gRPC port
# 11010 - REST port 
#expose-ports = [10735,11009,11010]
```

But the gPRC and REST port numbers are wrong. The real ports are 10009 and 10010. So What do you think of this problem? I believe we should not give the user wrong information about nonexisted ports. Beause exposing 11009 and 11010 supposed to be failed at the monent.

BTW in a TOML file, arrays should always contain elements with the same type. This is a limitation that we cannot get over it with xud-docker if we insist using TOML as our configuration file format.

Comment (sangaman) [reply images/utils/launcher/config/mainnet.conf]: Right, I don't see a huge problem with strings in the `expose-ports` array, and as Yang Yang says we can't mix plain numbers with strings. I don't see a workaround right now besides the mapping, and I think if we stick to it a line explaining how the mapping works (which side is the port exposed to the host machine, which side is the port inside the container) would be good.

**Pull request merged**